### PR TITLE
Websocket input plugin

### PR DESCRIPTION
**Pull request by srebhan**
### Required for all PRs:

- [X] Associated README.md updated.
- [X] Has appropriate unit tests.

This input plugin supports gathering data from a websocket endpoint. It has two operation modes, one event-based mode waiting for the server to push data and one actively gathering the data in regular intervals.

### CODE REVIEW

**Comment by ssoroka** [comment]
should there be a parameter for which mode is expected? Could you have both an `event` and `poll` mode with a trigger body (event using it just to trigger once?)

**Comment by srebhan** [comment]
@ssoroka currently `handshake_body` is sent after establishing connection (once). The `trigger_body` is sent on each gather cycle, so currently I use `trigger_body` as switch between using `Gather()` (sampling) and `Start()` with `listen()` (event-based) for acquisition. The reason is that I have two use-cases, one where we need to send a handshake-message and then receive data upon change of data and another one, where we need to send a handshake-message (like registering the client) and then a trigger-message to trigger sending of data.
We even might cover a third use-case if we allow multiple handshake-messages...

What do you suggest?

**Comment by srebhan** [comment]
!retry-failed

**Review by jessingrass — COMMENTED** [review]
I haven't looked at settings yet, but would appreciate feedback on the library choice when you have time!

File: docs/LICENSE_OF_DEPENDENCIES.md [hunk 1]
```diff
@@ -192,6 +192,7 @@ following works:
 - modernc.org/libc [BSD 3-Clause "New" or "Revised" License](https://gitlab.com/cznic/libc/-/blob/master/LICENSE)
 - modernc.org/memory [BSD 3-Clause "New" or "Revised" License](https://gitlab.com/cznic/memory/-/blob/master/LICENSE)
 - modernc.org/sqlite [BSD 3-Clause "New" or "Revised" License](https://gitlab.com/cznic/sqlite/-/blob/master/LICENSE)
+- nhooyr.io/websocket [MIT License](https://github.com/nhooyr/websocket/blob/master/LICENSE.txt)
```
[the remark this answers is not in the document]
Comment (srebhan) [reply docs/LICENSE_OF_DEPENDENCIES.md:195]: Hey @jessingrass, I've chosen this one as it seemed simpler for my use-case and didn't notice we already have gorilla. If you want me to change I can try...

**Comment by reimda** [comment]
Hi Sven, I just got a chance to look at this. I like it and have some questions. How do you intend to use this input? Are you going to write a custom websocket server or use an existing service? Many websocket servers require authentication and have a custom request protocol that might be too complex for the trigger_body mechanism.

I also think it would be good to switch this to gorilla/websocket so telegraf uses only one websocket library. I don't see any other plugins that use gorilla/websocket currently but some of our module dependencies use it, and the recent websocket output uses it.

**Comment by srebhan** [comment]
@reimda I use it for an existing service that can be handled with the means provided. Another important use-case is to connect two telegrafs in restricted (NATed) networks. Using the new websocket output plugin, we can connect two instances in setups where connection restrictions are in place.
However, I agree that this cannot cover arbitrary protocols, but it will be sufficient for the two use-cases above.

Currently, converting this to gorilla is of low priority here due to limited time. So if anybody (looks at @FZambia ;-)) wants to step up on it, any help is appreciated. :-)

**Comment by reimda** [comment]
> I use it for an existing service that can be handled with the means provided. Another important use-case is to connect two telegrafs in restricted (NATed) networks. Using the new websocket output plugin, we can connect two instances where an "inner" telegraf has to establish connects to an "outer" telegraf.

It would be useful to add some usage examples to the readme. Maybe these would be good examples to add if you get a chance. Thanks!

**Comment by srebhan** [comment]
Will close this as there seems to be no great interest in this PR. If anyone wants to continue here, please feel free to take the code and work on it.

**Pull request closed without merging**